url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/Networking/Server.java

```diff
@@ -0,0 +1,502 @@
+			sendDataToPartsRobot(req);
+		} else if (target.contains(Constants.NEST_TARGET)) {
+			sendDataToNest(req);
+		} else if (target.contains(Constants.CAMERA_TARGET)) {
+			sendDataToCamera(req);
+		} else if (target.contains(Constants.LANE_TARGET)) {
+			sendDataToLane(req);
+		} else if (target.contains(Constants.FEEDER_TARGET)) {
+			sendDataToLane(req);
+		} else if (target.contains(Constants.STAND_TARGET)) {
+			sendDataToStand(req);
+		} else if (target.contains(Constants.ALL_TARGET)) {
+			sendDataToGUIManagers(req);
+		} else if (target.contains(Constants.GANTRY_ROBOT_TARGET)) {
+			sendDataToGantry(req);
+		} else if (target.contains(Constants.MESSAGING_BOX_TARGET)) {
+			sendDataToMessage(req);
+		}
+	}
+
+	/**
+	 * This method sends the request to all managers with a GUI.
+	 * @param req
+	 */
+	private void sendDataToGUIManagers(Request req) {
+		// TODO - Remove - Temp hack to run FPM without running other clients.
+		if (factProdMngrWriter != null) {
+			factProdMngrWriter.sendData(req);
+		}
+
+		if (kitMngrWriter != null) {
+			kitMngrWriter.sendData(req);
+		}
+
+		if (partsMngrWriter != null) {
+			partsMngrWriter.sendData(req);
+		}
+	}
+
+	private void sendDataToConveyor(Request req) {
+		factProdMngrWriter.sendData(req);
+		kitRobotMngrWriter.sendData(req);
+		kitAssemblyMngrWriter.sendData(req);
+	}
+
+	private void sendDataToKitRobot(Request req) {
+		factProdMngrWriter.sendData(req);
+		kitRobotMngrWriter.sendData(req);
+		kitAssemblyMngrWriter.sendData(req);
+	}
+
+	private void sendDataToPartsRobot(Request req) {
+		factProdMngrWriter.sendData(req);
+		partsRobotMngrWriter.sendData(req);
+		kitAssemblyMngrWriter.sendData(req);
+	}
+
+	private void sendDataToStand(Request req) {
+		factProdMngrWriter.sendData(req);
+		kitRobotMngrWriter.sendData(req);
+		kitAssemblyMngrWriter.sendData(req);
+	}
+	private void sendDataToNest(Request req) {
+		factProdMngrWriter.sendData(req);
+		kitAssemblyMngrWriter.sendData(req);
+		laneMngrWriter.sendData(req);
+	}
+
+	private void sendDataToCamera(Request req) {
+		factProdMngrWriter.sendData(req);
+		partsRobotMngrWriter.sendData(req);
+		kitAssemblyMngrWriter.sendData(req);
+		laneMngrWriter.sendData(req);
+	}
+
+	private void sendDataToLane(Request req) {
+		factProdMngrWriter.sendData(req);
+		laneMngrWriter.sendData(req);
+		gantryRobotMngrWriter.sendData(req);
+	}
+
+	private void sendDataToGantry(Request req) {
+		factProdMngrWriter.sendData(req);
+		laneMngrWriter.sendData(req);
+		gantryRobotMngrWriter.sendData(req);
+	}
+
+	public void sendDataToMessage(Request req) {
+		factProdMngrWriter.sendData(req);
+	}
+
+	private void addShutdownHook() {
+		Thread hook = new Thread(new Runnable() {
+			@Override
+			public void run() {
+				saveFCSData();
+			}
+		});
+		Runtime.getRuntime().addShutdownHook(hook);
+	}
+
+	private void saveFCSData() {
+		try {
+			// save kit configs to KitConfigBackup.sav
+			FileOutputStream saveKitConfigs = new FileOutputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *saveKitConfigs*. The resource is referenced at lines 473, 475. The resource is closed at line 485. There are other execution paths that do not contain closure statements, for example, when FileOutputStream constructor throws an exception. Ensure *saveKitConfigs* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).